contracts/modules/commons/ModuleIPFS.sol

```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.17;
+
+import "./ModuleSelfAuth.sol";
+import "./ModuleStorage.sol";
+
+import "../../utils/LibString.sol";
+
+
+contract ModuleIPFS is ModuleSelfAuth {
+  event IPFSRootUpdated(bytes32 _hash);
+
+  //                       IPFS_ROOT_KEY = keccak256("sequence.ipfs.root")
+  bytes32 private constant IPFS_ROOT_KEY = bytes32(0x0eecac93ced8722d209199364cda3bc33da3bc3a23daef6be49ebd780511d033);
+
+  function ipfsRootBytes32() public view returns (bytes32) {
+    return ModuleStorage.readBytes32(IPFS_ROOT_KEY);
+  }
+
+  function ipfsRoot() public view returns (string memory) {
+    return string(
+      abi.encodePacked(
+        "ipfs://",
+        LibString.prefixBase32(
+          LibString.bytesToBase32(
+            abi.encodePacked(
+              hex'01701220',
```

Comment:
Where's this magic number coming from?

Reply:
This a pre-defined ipfs prefix.
See: https://docs.ipfs.tech/concepts/hashing/#important-hash-characteristics

The prefix we are using is:

- `01`: cidv1
- `70`: dag-pb
- `1220` (multihash)
- - `12` sha2-256
- - `20` 256 bits

The idea is that if we always encode the hash in the same way then we don't need to define a prefix per-wallet, so we can store the IPFS hash using a single storage slot.